csrc/root_domain_map.cpp

```diff
@@ -447,12 +447,39 @@ bool UnmappableReductionDomains::isReductionOutputMapped(
               input_keys.end();
         });
     // None of the consumer domains is used for the reduction
-    // domain. They should be safe with respect to this reduction
-    // domain
+    // domain.
+    if (it == consumer_domains.end()) {
+      // Further check if any consumer domain is mapped with the reduction input
+      // domains
+      it = std::find_if(
+          consumer_domains.begin(),
+          consumer_domains.end(),
+          [&](const auto& consumer_domain) {
+            return std::any_of(
+                input_keys.begin(),
+                input_keys.end(),
+                [&](const auto& input_key) {
+                  // check if input_key is concretized
+                  if (input_key.id()->isBroadcast()) {
+                    if (!root_map.isConcretized(
+                            input_key.td(), input_key.id())) {
+                      return false;
```

Comment:
Please add why this case should be ignored.

Reply:
added.
```
                  // canMap check requires concretized IDs for broadcast domain.
                  // For example, in softmax there are two consecutive
                  // reductions, one of the inputs to the 2nd reduction has a
                  // broadcast domain and it is not concretized until its
                  // consumer is visited where it will be concretized to its
                  // consumer's non-broadcast domain.
```

Reply:
That is correct, but that means the broadcast ID is not yet processed. Not just such broadcast domains, but there's generally no guarantee that reduction input domains are already processed. If not, `canMap` just returns false. Is the analysis still valid?

Reply:
The analaysis can only make a decision based on the current visited exprs. Let's still use `tv11` and its two consumers `tv12` and `tv13` as the example. If we are trying to do this mapping before `tv2 --> {tv13, tv14}` is visited. (This is similar to what you mensioned, where we didn't notice one of the consumer can be mapped with a reduction input) We will find it is OK to map  `tv11` with its two consumers.

Then the analysis continues,  it later finds a tensor `tv2` is an input to reduction domain and also used by `tv13`, this means we can't map `tv2` with its consumers. and `tv2` becomes a persistent tensor.

The generated code may be different if the visiting order is different. This is actually an example I put in the [doc](https://docs.google.com/document/d/1tZoJKmyX6EFf1OZipaIg2Ts3fm7LN6BuuyfQa94_oQg/edit#heading=h.5g2dpdnbafwu).

Reply:
if we want to guarantee the commutative law in fusion definition, e.g. `auto tv6 = add(tv1, tv5);` is always equivalent to ` auto tv6 = add(tv5, tv1);`. Looks like there is something needs to be done.

Reply:
> The analaysis can only make a decision based on the current visited exprs. Let's still use `tv11` and its two consumers `tv12` and `tv13` as the example. If we are trying to do this mapping before `tv2 --> {tv13, tv14}` is visited. (This is similar to what you mensioned, where we didn't notice one of the consumer can be mapped with a reduction input) We will find it is OK to map `tv11` with its two consumers.
> 
> Then the analysis continues, it later finds a tensor `tv2` is an input to reduction domain and also used by `tv13`, this means we can't map `tv2` with its consumers. and `tv2` becomes a persistent tensor.
> 
> The generated code may be different if the visiting order is different. This is actually an example I put in the [doc](https://docs.google.com/document/d/1tZoJKmyX6EFf1OZipaIg2Ts3fm7LN6BuuyfQa94_oQg/edit#heading=h.5g2dpdnbafwu).

That's correct. Please add that to the code as a comment.